packages/flutter_tools/test/general.shard/ios/ios_device_start_prebuilt_test.dart

```diff
@@ -0,0 +1,400 @@
+    });
+    when(MDnsObservatoryDiscovery.instance.getObservatoryUri(any, any, usesIpv6: anyNamed('usesIpv6')))
+      .thenAnswer((Invocation invocation) async => null);
+
+    final LaunchResult launchResult = await device.startApp(iosApp,
+      prebuiltApplication: true,
+      debuggingOptions: DebuggingOptions.enabled(BuildInfo.debug),
+      platformArgs: <String, dynamic>{},
+    );
+
+    expect(launchResult.started, true);
+    expect(launchResult.hasObservatory, true);
+    verify(globals.flutterUsage.sendEvent('ios-handshake', 'mdns-failure')).called(1);
+    verify(globals.flutterUsage.sendEvent('ios-handshake', 'fallback-success')).called(1);
+    expect(await device.stopApp(iosApp), false);
+  }, overrides: <Type, Generator>{
+    Usage: () => MockUsage(),
+    MDnsObservatoryDiscovery: () => MockMDnsObservatoryDiscovery(),
+  });
+
+  // Still uses context for analytics and mDNS.
+  testUsingContext('IOSDevice.startApp fails in debug mode when mDNS fails and '
+    'when Observatory URI is malformed', () async {
+    final FileSystem fileSystem = MemoryFileSystem.test();
+    final FakeProcessManager processManager = FakeProcessManager.list(<FakeCommand>[
+      kDeployCommand,
+      kLaunchDebugCommand,
+    ]);
+    final IOSDevice device = setUpIOSDevice(
+      processManager: processManager,
+      fileSystem: fileSystem,
+    );
+    final IOSApp iosApp = PrebuiltIOSApp(
+      projectBundleId: 'app',
+      bundleName: 'Runner',
+      bundleDir: fileSystem.currentDirectory,
+    );
+    final FakeDeviceLogReader deviceLogReader = FakeDeviceLogReader();
+
+    device.portForwarder = const NoOpDevicePortForwarder();
+    device.setLogReader(iosApp, deviceLogReader);
+
+    // Now that the reader is used, start writing messages to it.
+    Timer.run(() {
+      deviceLogReader.addLine('Foo');
+      deviceLogReader.addLine('Observatory listening on http:/:/127.0.0.1:456');
+    });
+    when(MDnsObservatoryDiscovery.instance.getObservatoryUri(any, any, usesIpv6: anyNamed('usesIpv6')))
+      .thenAnswer((Invocation invocation) async => null);
+
+    final LaunchResult launchResult = await device.startApp(iosApp,
+      prebuiltApplication: true,
+      debuggingOptions: DebuggingOptions.enabled(BuildInfo.debug),
+      platformArgs: <String, dynamic>{},
+    );
+
+    expect(launchResult.started, false);
+    expect(launchResult.hasObservatory, false);
+    verify(globals.flutterUsage.sendEvent(
+      'ios-handshake',
+      'failure-other',
+      label: anyNamed('label'),
+      value: anyNamed('value'),
+    )).called(1);
+    verify(globals.flutterUsage.sendEvent('ios-handshake', 'mdns-failure')).called(1);
+    verify(globals.flutterUsage.sendEvent('ios-handshake', 'fallback-failure')).called(1);
+    }, overrides: <Type, Generator>{
+      MDnsObservatoryDiscovery: () => MockMDnsObservatoryDiscovery(),
+      Usage: () => MockUsage(),
+    });
+
+  // Still uses context for analytics and mDNS.
```

Comment:
If it's using them, should we be overriding them?

Reply:
Updated the comment, and provided a Usage